openml/extensions/extension_interface.py

```diff
@@ -0,0 +1,90 @@
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional, Tuple
+
+from openml import OpenMLTask, OpenMLFlow
+from openml.runs.trace import OpenMLTraceIteration, OpenMLRunTrace
+
+
+class Extension(ABC):
+
+    """Defines the interface to connect machine learning libraries to OpenML-Python.
+    """
+
+    ################################################################################################
+    # Abstract methods for flow serialization and de-serialization
+
+    @abstractmethod
+    def flow_to_model(self, flow: OpenMLFlow) -> Any:
+        pass
+
+    @abstractmethod
+    def model_to_flow(self, model: Any) -> OpenMLFlow:
+        pass
+
+    @abstractmethod
+    def flow_to_parameters(self, model: Any) -> List:
+        pass
+
+    @abstractmethod
+    def is_estimator(self, model: Any) -> bool:
+        pass
+
+    @abstractmethod
+    def seed_model(self, model: Any, seed: Optional[int]) -> Any:
+        pass
+
+    @abstractmethod
+    def run_model_on_fold(
```

Comment:
I'm undecided whether this function should be in this interface or in the `openml-python` connector (making use of call_back functions `model_predict` and `model_fit`).

Reply:
I was undecided too until I figured out that there's more interaction going on here:

* the model needs to copied/cloned which is specific to each library
* there's some code specific to pipelines included to figure out which classes are actually used by the scikit-learn estimator
* we would need `model_predict()` and `model_predict_proba()`
* as soon as we start adding timing for parallelized grid search and random search, this will require library specific code as well.

Reply:
I think I agree. Still I would consider making this an easier function. For example, looking at its signature, it currently has a wide range of arguments. I would almost guess that we don't need `task`, `rep_no`, `fold_no` and `sample_no` since we can provide `X` and `y`, and we should probably be a bit more clear on the `tuple` that is returned (I would propose: a numpy array with predictions corresponding to the `y` variable and a dict containing the local measures).

Reply:
Great idea regarding `X` and `y`, I'll take it into account. I'll be more specific regarding the tuple once I start working on the documentation.

Reply:
I just tried to apply this simplification and stumbled upon a few issues which suggest that this should be implemented separately in a future pull request:

1. The task object is actually required so that the run function can decide on the call args for the fit function.
2. Simplifying the return value probably requires adding functionality to the `Task` class on how to transform predictions in valid content for uploading. I would like to avoid adding this in this PR as this will require larger refactoring regarding the way the prediction files are created.

Would that be okay for you for now?

Reply:
Thanks for pointing me back to this thread @mfeurer, I had missed the follow-up discussion. I am in favor of continuing this PR as is, and then opening an issue so that we remind ourselves to refactor this afterwards 👍